File: tests/Feature/HorizonteServiceTest.php
<?php

namespace Tests\Feature;

use App\Models\Account;
use App\Models\Divida;
use App\Models\PerfilFinanceiro;
use App\Models\User;
use App\Services\HorizonteService;
use Carbon\CarbonImmutable;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Tests\TestCase;

/**
 * Item 2: o horizonte que fala.
 *
 * O risco que estes testes guardam é uma parcela sumir da projeção. Um
 * vencimento dia 31 não existe em fevereiro; se o cálculo exigir dia
 * exato, a parcela desaparece justo no mês mais curto — e o app mostra
 * um mês folgado que não existe.
 */
class HorizonteServiceTest extends TestCase
{
    use RefreshDatabase;

    private function service(): HorizonteService
    {
        return app(HorizonteService::class);
    }

    private function usuario(float $saldo = 0, array $perfil = [], array $dividas = []): User
    {
        $user = User::factory()->create();

        Account::create([
            'user_id' => $user->id,
            'name' => 'Conta',
            'type' => 'bank',
            'initial_balance' => $saldo,
            'current_balance' => $saldo,
        ]);

        if ($perfil !== []) {
            PerfilFinanceiro::factory()->for($user)->create($perfil);
        }

        foreach ($dividas as $d) {
            Divida::factory()->for($user)->create($d);
        }

        return $user;
    }

    public function test_cada_linha_tem_nome_e_valor(): void
    {
        $user = $this->usuario(5000,
            ['renda_mensal' => 3000, 'dia_renda' => 5, 'gasto_diario_estimado' => 0,
             'contas_fixas_estimadas' => 0],
            [['nome' => 'Nubank', 'valor_parcela' => 1534.06,
              'dia_vencimento' => 4, 'parcelas_restantes' => 4]],
        );

        $h = $this->service()->mes($user->id, CarbonImmutable::create(2026, 9, 1));

        $dia4 = collect($h['dias'])->firstWhere('dia', 4);
        $this->assertSame('parcela do Nubank', $dia4['lancamentos'][0]['nome']);
        $this->assertSame(1534.06, $dia4['lancamentos'][0]['valor']);
        $this->assertSame('saida', $dia4['lancamentos'][0]['tipo']);

        $dia5 = collect($h['dias'])->firstWhere('dia', 5);
        $this->assertSame('salário', $dia5['lancamentos'][0]['nome']);
        $this->assertSame('entrada', $dia5['lancamentos'][0]['tipo']);
    }

    public function test_explica_quando_a_parcela_cai_antes_do_salario(): void
    {
        // O caso que custou R$ 1.674 ao Gabriel em 12 meses.
        $user = $this->usuario(100,
            ['renda_mensal' => 3000, 'dia_renda' => 6, 'gasto_diario_estimado' => 0,
             'contas_fixas_estimadas' => 0],
            [['nome' => 'Nubank', 'valor_parcela' => 1534.06,
              'dia_vencimento' => 4, 'parcelas_restantes' => 4]],
        );

        $h = $this->service()->mes($user->id, CarbonImmutable::create(2026, 9, 1));
        $dia4 = collect($h['dias'])->firstWhere('dia', 4);

        $this->assertSame('aperto', $dia4['estado']);
        $this->assertSame(
            'a parcela do Nubank cai antes do salário do dia 6',
            $dia4['motivo']
        );
    }

    public function test_dia_tranquilo_nao_tem_motivo(): void
    {
        $user = $this->usuario(10000,
            ['renda_mensal' => 3000, 'dia_renda' => 5, 'gasto_diario_estimado' => 0,
             'contas_fixas_estimadas' => 0],
        );

        $h = $this->service()->mes($user->id, CarbonImmutable::create(2026, 9, 1));

        $this->assertSame('tranquilo', $h['dias'][0]['estado']);
        $this->assertNull($h['dias'][0]['motivo']);
    }

    public function test_saldo_baixo_vira_atencao_antes_de_estourar(): void
    {
        $user = $this->usuario(50, ['renda_mensal' => 0, 'dia_renda' => null,
            'gasto_diario_estimado' => 0, 'contas_fixas_estimadas' => 0]);

        $h = $this->service()->mes($user->id, CarbonImmutable::create(2026, 9, 1));

        $this->assertSame('atencao', $h['dias'][0]['estado']);
        $this->assertSame('o saldo fica baixo neste dia', $h['dias'][0]['motivo']);
    }

    public function test_parcela_dia_31_nao_some_em_fevereiro(): void
    {
        $user = $this->usuario(10000,
            ['renda_mensal' => 0, 'dia_renda' => null, 'gasto_diario_estimado' => 0,
             'contas_fixas_estimadas' => 0],
            [['nome' => 'Crediário', 'valor_parcela' => 200,
              'dia_vencimento' => 31, 'parcelas_restantes' => 12]],
        );

        $fev = $this->service()->mes($user->id, CarbonImmutable::create(2026, 2, 1));

        $ultimo = collect($fev['dias'])->last();
        $this->assertSame(28, $ultimo['dia']);
        $this->assertSame('parcela do Crediário', $ultimo['lancamentos'][0]['nome']);

        // E aparece uma única vez no mês.
        $comLancamento = collect($fev['dias'])->filter(fn ($d) => $d['lancamentos'] !== []);
        $this->assertCount(1, $comLancamento);
    }

    public function test_toda_combinacao_de_vencimento_e_renda_aparece_uma_vez_por_mes(): void
    {
        // 24 meses × vencimento 1–31 × renda 1–31, como pede o handoff.
        // A garantia: cada lançamento aparece exatamente uma vez no mês,
        // nunca zero (some) nem duas (duplica).
        $s = $this->service();

        foreach ([1, 4, 15, 28, 29, 30, 31] as $venc) {
            foreach ([1, 5, 28, 31] as $diaRenda) {
                $user = $this->usuario(50000,
                    ['renda_mensal' => 3000, 'dia_renda' => $diaRenda,
                     'gasto_diario_estimado' => 0, 'contas_fixas_estimadas' => 0],
                    [['nome' => 'X', 'valor_parcela' => 100,
                      'dia_vencimento' => $venc, 'parcelas_restantes' => 36]],
                );

                for ($i = 0; $i < 24; $i++) {
                    $mes = CarbonImmutable::create(2026, 1, 1)->addMonthsNoOverflow($i);
                    $h = $s->mes($user->id, $mes);

                    $parcelas = collect($h['dias'])->flatMap(fn ($d) => $d['lancamentos'])
                        ->where('nome', 'parcela do X');
                    $salarios = collect($h['dias'])->flatMap(fn ($d) => $d['lancamentos'])
                        ->where('nome', 'salário');

                    $this->assertCount(1, $parcelas,
                        "vencimento {$venc} em {$mes->format('Y-m')}");
                    $this->assertCount(1, $salarios,
                        "renda dia {$diaRenda} em {$mes->format('Y-m')}");
                }
            }
        }
    }

    public function test_horizonte_de_12_meses_mostra_a_divida_acabando(): void
    {
        $user = $this->usuario(0,
            ['renda_mensal' => 3000, 'dia_renda' => 5, 'gasto_diario_estimado' => 0,
             'contas_fixas_estimadas' => 1000],
            [['nome' => 'Nubank', 'valor_parcela' => 1500,
              'dia_vencimento' => 4, 'parcelas_restantes' => 4]],
        );

        $doze = $this->service()->doze($user->id, CarbonImmutable::create(2026, 9, 1));

        $this->assertCount(12, $doze);
        $this->assertSame('set/26', $doze[0]['label']);

        // Enquanto há parcela, o mês pesa 1500 a mais.
        $this->assertTrue($doze[0]['tem_divida']);
        $this->assertSame(1500.0, $doze[0]['parcelas']);
        $this->assertSame(500.0, $doze[0]['saldo_final']);

        // A 4ª parcela é a última: de dezembro em diante não pesa mais.
        $this->assertTrue($doze[3]['tem_divida']);
        $this->assertFalse($doze[4]['tem_divida']);
        $this->assertSame(0.0, $doze[4]['parcelas']);

        // E o saldo volta a subir 2000 por mês.
        $this->assertSame($doze[4]['saldo_final'] + 2000.0, $doze[5]['saldo_final']);
    }

    public function test_mes_negativo_no_horizonte_fica_em_aperto(): void
    {
        $user = $this->usuario(0,
            ['renda_mensal' => 2000, 'dia_renda' => 5, 'gasto_diario_estimado' => 0,
             'contas_fixas_estimadas' => 900],
            [['nome' => 'Nubank', 'valor_parcela' => 1534,
              'dia_vencimento' => 4, 'parcelas_restantes' => 6]],
        );

        $doze = $this->service()->doze($user->id, CarbonImmutable::create(2026, 9, 1));

        $this->assertSame('aperto', $doze[0]['estado']);
        $this->assertSame(-434.0, $doze[0]['saldo_final']);
    }

    public function test_marca_o_primeiro_dia_apertado_do_mes(): void
    {
        $user = $this->usuario(100,
            ['renda_mensal' => 3000, 'dia_renda' => 20, 'gasto_diario_estimado' => 0,
             'contas_fixas_estimadas' => 0],
            [['nome' => 'Nubank', 'valor_parcela' => 500,
              'dia_vencimento' => 10, 'parcelas_restantes' => 4]],
        );

        $h = $this->service()->mes($user->id, CarbonImmutable::create(2026, 9, 1));

        $this->assertSame('2026-09-10', $h['primeiro_dia_apertado']);
    }

    public function test_gasto_diario_consome_o_saldo(): void
    {
        $user = $this->usuario(1000,
            ['renda_mensal' => 0, 'dia_renda' => null, 'gasto_diario_estimado' => 10,
             'contas_fixas_estimadas' => 0],
        );

        $h = $this->service()->mes($user->id, CarbonImmutable::create(2026, 9, 1));

        $this->assertSame(990.0, $h['dias'][0]['saldo']);
        $this->assertSame(980.0, $h['dias'][1]['saldo']);
        // 30 dias × 10 = 300
        $this->assertSame(700.0, $h['saldo_final']);
    }

    public function test_divida_quitada_nao_entra_no_horizonte(): void
    {
        $user = $this->usuario(1000, ['renda_mensal' => 0, 'dia_renda' => null,
            'gasto_diario_estimado' => 0, 'contas_fixas_estimadas' => 0]);
        Divida::factory()->for($user)->quitada()->create([
            'nome' => 'Paga', 'valor_parcela' => 900, 'dia_vencimento' => 10,
        ]);

        $h = $this->service()->mes($user->id, CarbonImmutable::create(2026, 9, 1));

        $this->assertSame(1000.0, $h['saldo_final']);
        $comLancamento = collect($h['dias'])->filter(fn ($d) => $d['lancamentos'] !== []);
        $this->assertCount(0, $comLancamento);
    }

    public function test_usuario_sem_perfil_nao_quebra(): void
    {
        $user = $this->usuario(0);

        $h = $this->service()->mes($user->id, CarbonImmutable::create(2026, 9, 1));
        $doze = $this->service()->doze($user->id, CarbonImmutable::create(2026, 9, 1));

        $this->assertCount(30, $h['dias']);
        $this->assertCount(12, $doze);
    }
}

File: test/horizonte_service_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:kitamo/design/cores.dart';
import 'package:kitamo/models/divida.dart';
import 'package:kitamo/models/perfil_financeiro.dart';
import 'package:kitamo/services/horizonte_service.dart';

/// Fase 3: o horizonte.
///
/// O risco que estes testes guardam é uma parcela sumir da projeção. Um
/// vencimento dia 31 não existe em fevereiro; se o cálculo exigir dia
/// exato, a parcela desaparece justo no mês mais curto — e o app mostra um
/// mês folgado que não existe.
void main() {
  const service = HorizonteService();

  Divida divida({
    String nome = 'Nubank',
    double parcela = 1534.06,
    int vencimento = 4,
    int restantes = 4,
    int total = 10,
  }) =>
      Divida(
        id: nome,
        nome: nome,
        saldoAtual: 0,
        valorParcela: parcela,
        diaVencimento: vencimento,
        parcelasRestantes: restantes,
        parcelasTotal: total,
      );

  group('mês', () {
    test('cada linha tem nome e valor', () {
      final m = service.mes(
        perfil: const PerfilFinanceiro(rendaMensal: 3000, diaRenda: 5),
        dividas: [divida()],
        saldoInicial: 5000,
        referencia: DateTime(2026, 9, 1),
      );

      final dia4 = m.dias.firstWhere((d) => d.dia == 4);
      expect(dia4.lancamentos.first.nome, 'parcela do Nubank');
      expect(dia4.lancamentos.first.valor, 1534.06);
      expect(dia4.lancamentos.first.entrada, isFalse);

      final dia5 = m.dias.firstWhere((d) => d.dia == 5);
      expect(dia5.lancamentos.first.nome, 'salário');
      expect(dia5.lancamentos.first.entrada, isTrue);
    });

    test('explica quando a parcela cai antes do salário', () {
      final m = service.mes(
        perfil: const PerfilFinanceiro(rendaMensal: 3000, diaRenda: 6),
        dividas: [divida()],
        saldoInicial: 100,
        referencia: DateTime(2026, 9, 1),
      );

      final dia4 = m.dias.firstWhere((d) => d.dia == 4);
      expect(dia4.estado, EstadoFinanceiro.aperto);
      expect(dia4.motivo, 'a parcela do Nubank cai antes do salário do dia 6');
    });

    test('dia tranquilo não tem motivo', () {
      final m = service.mes(
        perfil: const PerfilFinanceiro(rendaMensal: 3000, diaRenda: 5),
        dividas: [],
        saldoInicial: 10000,
        referencia: DateTime(2026, 9, 1),
      );

      expect(m.dias.first.estado, EstadoFinanceiro.tranquilo);
      expect(m.dias.first.motivo, isNull);
    });

    test('saldo baixo vira atenção antes de estourar', () {
      final m = service.mes(
        perfil: const PerfilFinanceiro(),
        dividas: [],
        saldoInicial: 50,
        referencia: DateTime(2026, 9, 1),
      );

      expect(m.dias.first.estado, EstadoFinanceiro.atencao);
      expect(m.dias.first.motivo, 'o saldo fica baixo neste dia');
    });

    test('parcela dia 31 não some em fevereiro', () {
      final m = service.mes(
        perfil: const PerfilFinanceiro(),
        dividas: [divida(nome: 'Crediário', parcela: 200, vencimento: 31, restantes: 12)],
        saldoInicial: 10000,
        referencia: DateTime(2026, 2, 1),
      );

      final comLancamento = m.dias.where((d) => d.lancamentos.isNotEmpty);
      expect(comLancamento, hasLength(1));
      expect(comLancamento.first.dia, 28);
    });

    test('gasto diário consome o saldo', () {
      final m = service.mes(
        perfil: const PerfilFinanceiro(gastoDiarioEstimado: 10),
        dividas: [],
        saldoInicial: 1000,
        referencia: DateTime(2026, 9, 1),
      );

      expect(m.dias[0].saldo, 990);
      expect(m.dias[1].saldo, 980);
      expect(m.saldoFinal, 700); // 30 dias × 10
    });

    test('marca o primeiro dia apertado', () {
      final m = service.mes(
        perfil: const PerfilFinanceiro(rendaMensal: 3000, diaRenda: 20),
        dividas: [divida(parcela: 500, vencimento: 10)],
        saldoInicial: 100,
        referencia: DateTime(2026, 9, 1),
      );

      expect(m.primeiroDiaApertado, DateTime(2026, 9, 10));
    });

    test('dívida quitada não entra', () {
      final quitada = Divida(
        id: 'q',
        nome: 'Paga',
        saldoAtual: 0,
        valorParcela: 900,
        diaVencimento: 10,
        parcelasRestantes: 0,
        parcelasTotal: 3,
        quitadaEm: DateTime(2026, 1, 1),
      );

      final m = service.mes(
        perfil: const PerfilFinanceiro(),
        dividas: [quitada],
        saldoInicial: 1000,
        referencia: DateTime(2026, 9, 1),
      );

      expect(m.saldoFinal, 1000);
      expect(m.dias.where((d) => d.lancamentos.isNotEmpty), isEmpty);
    });

    test('sem perfil não quebra', () {
      final m = service.mes(
        perfil: null,
        dividas: [],
        saldoInicial: 0,
        referencia: DateTime(2026, 9, 1),
      );

      expect(m.dias, hasLength(30));
    });

    test('toda combinação de vencimento e renda aparece uma vez por mês', () {
      // 24 meses × 7 vencimentos × 4 dias de renda. A garantia: cada
      // lançamento aparece exatamente uma vez — nunca zero (some), nunca
      // duas (duplica).
      for (final venc in [1, 4, 15, 28, 29, 30, 31]) {
        for (final diaRenda in [1, 5, 28, 31]) {
          for (var i = 0; i < 24; i++) {
            final ref = DateTime(2026, 1 + i, 1);

            final m = service.mes(
              perfil: PerfilFinanceiro(rendaMensal: 3000, diaRenda: diaRenda),
              dividas: [divida(nome: 'X', parcela: 100, vencimento: venc, restantes: 36)],
              saldoInicial: 50000,
              referencia: ref,
            );

            final parcelas = m.dias
                .expand((d) => d.lancamentos)
                .where((l) => l.nome == 'parcela do X');
            final salarios = m.dias
                .expand((d) => d.lancamentos)
                .where((l) => l.nome == 'salário');

            expect(parcelas, hasLength(1),
                reason: 'vencimento $venc em ${ref.year}-${ref.month}');
            expect(salarios, hasLength(1),
                reason: 'renda dia $diaRenda em ${ref.year}-${ref.month}');
          }
        }
      }
    });
  });

  group('12 meses', () {
    test('mostra a dívida acabando', () {
      final doze = service.doze(
        perfil: const PerfilFinanceiro(
          rendaMensal: 3000,
          diaRenda: 5,
          contasFixasEstimadas: 1000,
        ),
        dividas: [divida(parcela: 1500, restantes: 4)],
        saldoInicial: 0,
        referencia: DateTime(2026, 9, 1),
      );

      expect(doze, hasLength(12));
      expect(doze[0].rotulo, 'set/26');
      expect(doze[0].temDivida, isTrue);
      expect(doze[0].saldoFinal, 500);

      // A 4ª parcela é a última: de janeiro em diante não pesa mais.
      expect(doze[3].temDivida, isTrue);
      expect(doze[4].temDivida, isFalse);

      // E o saldo volta a subir 2000 por mês.
      expect(doze[5].saldoFinal, doze[4].saldoFinal + 2000);
    });

    test('marca o mês da quitação', () {
      final doze = service.doze(
        perfil: const PerfilFinanceiro(rendaMensal: 3000, contasFixasEstimadas: 500),
        dividas: [
          divida(nome: 'A', parcela: 100, restantes: 2),
          divida(nome: 'B', parcela: 100, restantes: 5),
        ],
        saldoInicial: 0,
        referencia: DateTime(2026, 9, 1),
      );

      // A mais longa tem 5 parcelas: a última cai no índice 4.
      expect(doze[4].ehQuitacao, isTrue);
      expect(doze.where((m) => m.ehQuitacao), hasLength(1));
    });

    test('sem dívida nenhum mês é de quitação', () {
      final doze = service.doze(
        perfil: const PerfilFinanceiro(rendaMensal: 3000),
        dividas: [],
        saldoInicial: 0,
        referencia: DateTime(2026, 9, 1),
      );

      expect(doze.where((m) => m.ehQuitacao), isEmpty);
    });

    test('mês negativo fica em aperto', () {
      final doze = service.doze(
        perfil: const PerfilFinanceiro(rendaMensal: 2000, contasFixasEstimadas: 900),
        dividas: [divida(parcela: 1534, restantes: 6)],
        saldoInicial: 0,
        referencia: DateTime(2026, 9, 1),
      );

      expect(doze[0].estado, EstadoFinanceiro.aperto);
      expect(doze[0].saldoFinal, -434);
    });
  });
}
